% Chapter Template

\chapter{Introduction} % Main chapter title

% \label{Intro} % Change X to a consecutive number; for referencing this chapter elsewhere, use \ref{ChapterX}

\lhead{Chapter 1. \emph{Introduction}} % Change X to a consecutive number; this is for the header on each page - perhaps a shortened title

%----------------------------------------------------------------------------------------
%	SECTION 1
%----------------------------------------------------------------------------------------
\section{Problem Statement}
What's it + why important

History of problem + state of art

Summary of work + contribution

Discuss common use of HMMER

HMMER \citep{HMMER} is a free and commonly used software package for sequence analysis written by Sean Eddy. 
It is an open source implementation of HMM algorithms for use with protein databases.

One of its more widely used applications, \emph{hmmsearch} is to identify homologous protein. It does this with Viterbi algorithm by comparing a profile HMM to each protein sequence in a large database, evaluating the path that has the maximum probability of the HMM generating the sequence.
This search requires a computationally intensive procedure.

There has been a great deal of work on optimizing HMMer for traditional CPUs: JackHMMer[Wun et al., 2005] [Walters et al. 2006].

GPUs have been shown to provide very attractive compute and bandwidth resources in addition to CPUs.

MPI-HMMER [24] is a wellknown
and commonly used implementation.

J. P. Walters, B. Qudah, and V. Chaudhary. Accelerating
the HMMER Sequence Analysis Suite Using Conventional
Processors. In AINA ’06: Proceedings of the 20th
International Conference on Advanced Information Networking
and Applications, pages 289–294, Washington,
DC, USA, 2006. IEEE Computer Society.

John Paul Walters, Joseph Landman and Vipin Chaudhary 2006 Optimized ClusterEnabled HMMER Searches 
In book: Grid Computing for Bioinformatics and Computational Biology, pp.51 - 70
https://www.researchgate.net/publication/227987859_Optimized_ClusterEnabled_HMMER_Searches?ev=srch_pub&_sg=90v0%2FM%2FObKOe0lGb7vt630TN%2BCVRfFm%2FrSTeF%2FC%2FeNV905UE%2Bt8ZZo9urkZEMIXr_uMgA2HwKuq8e6BYfOc8Thl2%2Fxpk223W%2BesVp7mJaPT2KiUCj3VedL%2BXQqaxBJHnv

https://www.google.ca/url?sa=t&rct=j&q=&esrc=s&source=web&cd=4&cad=rja&uact=8&ved=0CFEQFjAD&url=http%3A%2F%2Fwww.researchgate.net%2Fpublication%2F227987859_Optimized_ClusterEnabled_HMMER_Searches%2Ffile%2F60b7d517e7501cc6aa.pdf&ei=5AqBU6b8HMS-8gGBj4HoDw&usg=AFQjCNFoTh5akZMbF3au8OlufUUGp864PA&sig2=hWFbM1CabJR4qwUUh8vmww


HMMER \citep{HMMER} is a set of applications that create a profile Hidden Markov Model (HMM) of a sequence family which can be utilized as a query against a sequence database to identify (and/or align) additional homologs of the sequence family\citep{Seq}. HMMER was developed by Sean Eddy at Washington University and has become one of the most widely used software tools for sequence homology.

HMMer [Eddy 2003a]  One of the more widely used algorithms,
hmmsearch, works as follows: a user provides an
HMM modeling a desired protein family and hmmsearch
processes each protein sequence in a large database, evaluating
the probability that the most likely path through the
query HMM could generate that database protein sequence.
This search requires a computationally intensive procedure,
known as the Viterbi [1967; 1973] algorithm. The search
could take hours or even days depending on the size of the
database, query model, and the processor used.

hmmsearch is widely used in the biology
research community.

There has been
a great deal of work on optimizing HMMer for traditional
CPUs [Lindahl 2005; Cofer and SGI. 2002].


\section{Research Contributions}

\section{Organization of thesis}

\subsection*{Typographical Conventions}
The following font conventions are used in this thesis:
\begin{itemize}
 \item {\fontfamily{phv}\fontseries{m}\selectfont Adobe Helvetica font}\\
 Used for code examples.
 \item {\fontfamily{phv}\fontseries{m}\selectfont \textsl{Adobe Helvetica slanted font}}\\
 Used for comments of code.
 \item {\fontfamily{pag}\selectfont Adobe AvantGarde font}\\
 Used for captions of table and figure.
\end{itemize}


%-----------------------------------
%	SUBSECTION 1
%-----------------------------------
\subsection{Subsection 1}


%-----------------------------------
%	SUBSECTION 2
%-----------------------------------

\subsection{Subsection 2}

%----------------------------------------------------------------------------------------
%	SECTION 2
%----------------------------------------------------------------------------------------

\section{Main Section 2}

